### ISSUE TITLE
test_verify_latest_copy_removenode_node: shared_ptr accessed on non-owner cpu

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylladb#11826]
Seen since https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-daily-debug/70/artifact/logs-full.debug.028/1666414874050_update_cluster_layout_tests.py%3A%3ATestUpdateClusterLayout%3A%3Atest_verify_latest_copy_removenode_node/node1.log
```
INFO  2022-10-22 05:00:30,950 [shard 0] init - Scylla version 5.2.0~dev-0.20221022.4aa0b1685265 with build-id fd5a735818da4f244d9031b8db0756796e2e9e2e starting ...

INFO  2022-10-22 05:01:10,981 [shard 0] repair - repair[7ae049ff-d68d-4fcb-bff5-13d4f9d28632]: sync data for keyspace=ks, status=started
INFO  2022-10-22 05:01:10,984 [shard 0] repair - repair[7ae049ff-d68d-4fcb-bff5-13d4f9d28632]: Started to repair 1 out of 1 tables in keyspace=ks, table=cf, table_id=12123130-51c6-11ed-900a-0cd4fa6c437d, repair_reason=removenode
ERROR 2022-10-22 05:01:10,985 [shard 1] seastar - shared_ptr accessed on non-owner cpu, at: 0x14f4a190 0x1d5d41ad 0x1d5d3e6d 0x1d5d6463 0x1d5d72f2 0x1c6611de 0x1c661d0a 0x150d1e55 0x150d1c65 0x1515df76 0x19e85dc3 0x19e85ad3 0x19e85a93 0x19e85a4b 0x19e85a03 0x19e858ce 0x19e84eb0 0x19e874c8 0x1c772a3e 0x1c778eae 0x1c77db12 0x1c86399b 0x1c862522 0x1c8623d2 0x1c861a22 0x1561ea7a 0x1c66618b /jenkins/workspace/scylla-master/dtest-daily-debug/scylla/.ccm/scylla-repository/4aa0b16852656ce11f13b78156fbf28cffbe8c1a/scylla/libreloc/libpthread.so.0+0x92a4 /jenkins/workspace/scylla-master/dtest-daily-debug/scylla/.ccm/scylla-repository/4aa0b16852656ce11f13b78156fbf28cffbe8c1a/scylla/libreloc/libc.so.6+0x100322
   --------
   seastar::smp_message_queue::async_work_item<seastar::sharded<repair_service>::invoke_on<repair_service::do_sync_data_using_repair(seastar::basic_sstring<char, unsigned int, 15u, true>, std::vector<nonwrapping_interval<dht::token>, std::allocator<nonwrapping_interval<dht::token> > >, std::unordered_map<nonwrapping_interval<dht::token>, repair_neighbors, std::hash<nonwrapping_interval<dht::token> >, std::equal_to<nonwrapping_interval<dht::token> >, std::allocator<std::pair<nonwrapping_interval<dht::token> const, repair_neighbors> > >, streaming::stream_reason, seastar::shared_ptr<node_ops_info>)::$_27::operator()()::{lambda(repair_service&)#1}, , seastar::future<void> >(unsigned int, seastar::smp_submit_to_options, repair_service::do_sync_data_using_repair(seastar::basic_sstring<char, unsigned int, 15u, true>, std::vector<nonwrapping_interval<dht::token>, std::allocator<nonwrapping_interval<dht::token> > >, std::unordered_map<nonwrapping_interval<dht::token>, repair_neighbors, std::hash<nonwrapping_interval<dht::token> >, std::equal_to<nonwrapping_interval<dht::token> >, std::allocator<std::pair<nonwrapping_interval<dht::token> const, repair_neighbors> > >, streaming::stream_reason, seastar::shared_ptr<node_ops_info>)::$_27::operator()()::{lambda(repair_service&)#1}&&)::{lambda()#1}>
Aborting on shard 1.
Backtrace:
  0x14f4a190
  0x1c8f28fd
  0x1c8f24fc
  0x1c753784
  0x1c77f68e
  0x1c86155c
  0x1c8617bd
  0x1c8615ea
  0x7f09a34b9a1f
  /jenkins/workspace/scylla-master/dtest-daily-debug/scylla/.ccm/scylla-repository/4aa0b16852656ce11f13b78156fbf28cffbe8c1a/scylla/libreloc/libc.so.6+0x3d2a1
  /jenkins/workspace/scylla-master/dtest-daily-debug/scylla/.ccm/scylla-repository/4aa0b16852656ce11f13b78156fbf28cffbe8c1a/scylla/libreloc/libc.so.6+0x268a3
  0x1c661d14
  0x150d1e55
  0x150d1c65
  0x1515df76
  0x19e85dc3
  0x19e85ad3
  0x19e85a93
  0x19e85a4b
  0x19e85a03
  0x19e858ce
  0x19e84eb0
  0x19e874c8
  0x1c772a3e
  0x1c778eae
  0x1c77db12
  0x1c86399b
  0x1c862522
  0x1c8623d2
  0x1c861a22
  0x1561ea7a
  0x1c66618b
  /jenkins/workspace/scylla-master/dtest-daily-debug/scylla/.ccm/scylla-repository/4aa0b16852656ce11f13b78156fbf28cffbe8c1a/scylla/libreloc/libpthread.so.0+0x92a4
  /jenkins/workspace/scylla-master/dtest-daily-debug/scylla/.ccm/scylla-repository/4aa0b16852656ce11f13b78156fbf28cffbe8c1a/scylla/libreloc/libc.so.6+0x100322
```

Decoded:
```
[Backtrace #0]
__interceptor_backtrace.part.0 at crtstuff.c:?
void seastar::backtrace<seastar::current_backtrace_tasklocal()::$_3>(seastar::current_backtrace_tasklocal()::$_3&&) at ./build/debug/seastar/./seastar/include/seastar/util/backtrace.hh:59
seastar::current_backtrace_tasklocal() at ./build/debug/seastar/./seastar/src/util/backtrace.cc:86
seastar::current_tasktrace() at ./build/debug/seastar/./seastar/src/util/backtrace.cc:137
seastar::current_backtrace() at ./build/debug/seastar/./seastar/src/util/backtrace.cc:170
log_error_and_backtrace(seastar::logger&, std::basic_string_view<char, std::char_traits<char> >) at ./build/debug/seastar/./seastar/src/core/on_internal_error.cc:37
seastar::on_fatal_internal_error(seastar::logger&, std::basic_string_view<char, std::char_traits<char> >) at ./build/debug/seastar/./seastar/src/core/on_internal_error.cc:67
seastar::debug_shared_ptr_counter_type::check() const at ././seastar/include/seastar/core/shared_ptr_debug_helper.hh:68
seastar::debug_shared_ptr_counter_type::operator++() at ././seastar/include/seastar/core/shared_ptr_debug_helper.hh:48
shared_ptr at ././seastar/include/seastar/core/shared_ptr.hh:512
operator() at ./repair/repair.cc:1300
seastar::future<void> std::__invoke_impl<seastar::future<void>, repair_service::do_sync_data_using_repair(seastar::basic_sstring<char, unsigned int, 15u, true>, std::vector<nonwrapping_interval<dht::token>, std::allocator<nonwrapping_interval<dht::token> > >, std::unordered_map<nonwrapping_interval<dht::token>, repair_neighbors, std::hash<nonwrapping_interval<dht::token> >, std::equal_to<nonwrapping_interval<dht::token> >, std::allocator<std::pair<nonwrapping_interval<dht::token> const, repair_neighbors> > >, streaming::stream_reason, seastar::shared_ptr<node_ops_info>)::$_27::operator()()::{lambda(repair_service&)#1}, repair_service&>(std::__invoke_other, repair_service::do_sync_data_using_repair(seastar::basic_sstring<char, unsigned int, 15u, true>, std::vector<nonwrapping_interval<dht::token>, std::allocator<nonwrapping_interval<dht::token> > >, std::unordered_map<nonwrapping_interval<dht::token>, repair_neighbors, std::hash<nonwrapping_interval<dht::token> >, std::equal_to<nonwrapping_interval<dht::token> >, std::allocator<std::pair<nonwrapping_interval<dht::token> const, repair_neighbors> > >, streaming::stream_reason, seastar::shared_ptr<node_ops_info>)::$_27::operator()()::{lambda(repair_service&)#1}&&, repair_service&) at /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/invoke.h:61
std::__invoke_result<repair_service::do_sync_data_using_repair(seastar::basic_sstring<char, unsigned int, 15u, true>, std::vector<nonwrapping_interval<dht::token>, std::allocator<nonwrapping_interval<dht::token> > >, std::unordered_map<nonwrapping_interval<dht::token>, repair_neighbors, std::hash<nonwrapping_interval<dht::token> >, std::equal_to<nonwrapping_interval<dht::token> >, std::allocator<std::pair<nonwrapping_interval<dht::token> const, repair_neighbors> > >, streaming::stream_reason, seastar::shared_ptr<node_ops_info>)::$_27::operator()()::{lambda(repair_service&)#1}, repair_service&>::type std::__invoke<repair_service::do_sync_data_using_repair(seastar::basic_sstring<char, unsigned int, 15u, true>, std::vector<nonwrapping_interval<dht::token>, std::allocator<nonwrapping_interval<dht::token> > >, std::unordered_map<nonwrapping_interval<dht::token>, repair_neighbors, std::hash<nonwrapping_interval<dht::token> >, std::equal_to<nonwrapping_interval<dht::token> >, std::allocator<std::pair<nonwrapping_interval<dht::token> const, repair_neighbors> > >, streaming::stream_reason, seastar::shared_ptr<node_ops_info>)::$_27::operator()()::{lambda(repair_service&)#1}, repair_service&>(std::__invoke_result&&, (repair_service::do_sync_data_using_repair(seastar::basic_sstring<char, unsigned int, 15u, true>, std::vector<nonwrapping_interval<dht::token>, std::allocator<nonwrapping_interval<dht::token> > >, std::unordered_map<nonwrapping_interval<dht::token>, repair_neighbors, std::hash<nonwrapping_interval<dht::token> >, std::equal_to<nonwrapping_interval<dht::token> >, std::allocator<std::pair<nonwrapping_interval<dht::token> const, repair_neighbors> > >, streaming::stream_reason, seastar::shared_ptr<node_ops_info>)::$_27::operator()()::{lambda(repair_service&)#1}&&)...) at /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/invoke.h:96
decltype(auto) std::__apply_impl<repair_service::do_sync_data_using_repair(seastar::basic_sstring<char, unsigned int, 15u, true>, std::vector<nonwrapping_interval<dht::token>, std::allocator<nonwrapping_interval<dht::token> > >, std::unordered_map<nonwrapping_interval<dht::token>, repair_neighbors, std::hash<nonwrapping_interval<dht::token> >, std::equal_to<nonwrapping_interval<dht::token> >, std::allocator<std::pair<nonwrapping_interval<dht::token> const, repair_neighbors> > >, streaming::stream_reason, seastar::shared_ptr<node_ops_info>)::$_27::operator()()::{lambda(repair_service&)#1}, std::tuple<repair_service&>, 0ul>(repair_service::do_sync_data_using_repair(seastar::basic_sstring<char, unsigned int, 15u, true>, std::vector<nonwrapping_interval<dht::token>, std::allocator<nonwrapping_interval<dht::token> > >, std::unordered_map<nonwrapping_interval<dht::token>, repair_neighbors, std::hash<nonwrapping_interval<dht::token> >, std::equal_to<nonwrapping_interval<dht::token> >, std::allocator<std::pair<nonwrapping_interval<dht::token> const, repair_neighbors> > >, streaming::stream_reason, seastar::shared_ptr<node_ops_info>)::$_27::operator()()::{lambda(repair_service&)#1}&&, std::tuple<repair_service&>&&, std::integer_sequence<unsigned long, 0ul>) at /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/tuple:1858
decltype(auto) std::apply<repair_service::do_sync_data_using_repair(seastar::basic_sstring<char, unsigned int, 15u, true>, std::vector<nonwrapping_interval<dht::token>, std::allocator<nonwrapping_interval<dht::token> > >, std::unordered_map<nonwrapping_interval<dht::token>, repair_neighbors, std::hash<nonwrapping_interval<dht::token> >, std::equal_to<nonwrapping_interval<dht::token> >, std::allocator<std::pair<nonwrapping_interval<dht::token> const, repair_neighbors> > >, streaming::stream_reason, seastar::shared_ptr<node_ops_info>)::$_27::operator()()::{lambda(repair_service&)#1}, std::tuple<repair_service&> >(repair_service::do_sync_data_using_repair(seastar::basic_sstring<char, unsigned int, 15u, true>, std::vector<nonwrapping_interval<dht::token>, std::allocator<nonwrapping_interval<dht::token> > >, std::unordered_map<nonwrapping_interval<dht::token>, repair_neighbors, std::hash<nonwrapping_interval<dht::token> >, std::equal_to<nonwrapping_interval<dht::token> >, std::allocator<std::pair<nonwrapping_interval<dht::token> const, repair_neighbors> > >, streaming::stream_reason, seastar::shared_ptr<node_ops_info>)::$_27::operator()()::{lambda(repair_service&)#1}&&, std::tuple<repair_service&>&&) at /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/tuple:1869
operator() at ././seastar/include/seastar/core/sharded.hh:457
_ZN7seastar8futurizeINS_6futureIvEEE6invokeIRZNS_7shardedI14repair_serviceE9invoke_onIZZNS6_25do_sync_data_using_repairENS_13basic_sstringIcjLj15ELb1EEESt6vectorI20nonwrapping_intervalIN3dht5tokenEESaISF_EESt13unordered_mapISF_16repair_neighborsSt4hashISF_ESt8equal_toISF_ESaISt4pairIKSF_SJ_EEEN9streaming13stream_reasonENS_10shared_ptrI13node_ops_infoEEEN4$_27clEvEUlRS6_E_JES2_EET1_jNS_21smp_submit_to_optionsEOT_DpOT0_EUlvE_JEEES2_S14_S17_ at ././seastar/include/seastar/core/future.hh:2149
seastar::smp_message_queue::async_work_item<seastar::sharded<repair_service>::invoke_on<repair_service::do_sync_data_using_repair(seastar::basic_sstring<char, unsigned int, 15u, true>, std::vector<nonwrapping_interval<dht::token>, std::allocator<nonwrapping_interval<dht::token> > >, std::unordered_map<nonwrapping_interval<dht::token>, repair_neighbors, std::hash<nonwrapping_interval<dht::token> >, std::equal_to<nonwrapping_interval<dht::token> >, std::allocator<std::pair<nonwrapping_interval<dht::token> const, repair_neighbors> > >, streaming::stream_reason, seastar::shared_ptr<node_ops_info>)::$_27::operator()()::{lambda(repair_service&)#1}, , seastar::future<void> >(unsigned int, seastar::smp_submit_to_options, repair_service::do_sync_data_using_repair(seastar::basic_sstring<char, unsigned int, 15u, true>, std::vector<nonwrapping_interval<dht::token>, std::allocator<nonwrapping_interval<dht::token> > >, std::unordered_map<nonwrapping_interval<dht::token>, repair_neighbors, std::hash<nonwrapping_interval<dht::token> >, std::equal_to<nonwrapping_interval<dht::token> >, std::allocator<std::pair<nonwrapping_interval<dht::token> const, repair_neighbors> > >, streaming::stream_reason, seastar::shared_ptr<node_ops_info>)::$_27::operator()()::{lambda(repair_service&)#1}&&)::{lambda()#1}>::run_and_dispose() at ././seastar/include/seastar/core/smp.hh:231
seastar::reactor::run_tasks(seastar::reactor::task_queue&) at ./build/debug/seastar/./seastar/src/core/reactor.cc:2329
seastar::reactor::run_some_tasks() at ./build/debug/seastar/./seastar/src/core/reactor.cc:2736
seastar::reactor::do_run() at ./build/debug/seastar/./seastar/src/core/reactor.cc:2905
operator() at ./build/debug/seastar/./seastar/src/core/reactor.cc:4133
void std::__invoke_impl<void, seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_90&>(std::__invoke_other, seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_90&) at /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/invoke.h:61
std::enable_if<is_invocable_r_v<void, seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_90&>, void>::type std::__invoke_r<void, seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_90&>(seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_90&) at /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/invoke.h:111
std::_Function_handler<void (), seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_90>::_M_invoke(std::_Any_data const&) at /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/std_function.h:290
std::function<void ()>::operator()() const at /usr/lib/gcc/x86_64-redhat-linux/11/../../../../include/c++/11/bits/std_function.h:590
seastar::posix_thread::start_routine(void*) at ./build/debug/seastar/./seastar/src/core/posix.cc:73
```

Looks like the offending change is 8231b4ec1ba42d4e86ee2fdd20362a0ae3b484bb
where https://github.com/scylladb/scylladb/blob/3dc7c338477394a8c948282f276c6865276a85c5/repair/repair.cc#L1300
copies the `node_ops_info->as` shared_ptr across shards

**Comment by mykaul** [comment]
@xemul - do we need a 5.1 backport?

**Comment by bhalevy** [comment]
> @xemul - do we need a 5.1 backport?

No, since the root cause, https://github.com/scylladb/scylladb/commit/8231b4ec1ba42d4e86ee2fdd20362a0ae3b484bb, was introduced and fixed in 5.2

**Comment by denesb** [comment]
All live versions have this, removing label.
